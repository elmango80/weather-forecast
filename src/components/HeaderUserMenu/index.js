import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { htmlIdGenerator } from "@elastic/eui/lib/services";

import { signOut } from "redux/actions/auth";

import {
  EuiAvatar,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPopover,
  EuiSpacer,
  EuiText,
  EuiHeaderSectionItemButton,
  EuiButtonEmpty,
} from "@elastic/eui";

export default function HeaderUserMenu() {
  const [isOpen, setIsOpen] = useState(false);
  const dispatch = useDispatch();
  const { displayName, photoURL } = useSelector((state) => state.auth);

  const id = htmlIdGenerator()();

  const onMenuButtonClick = () => {
    setIsOpen(!isOpen);
  };

  const handleClosePopover = () => {
    setIsOpen(false);
  };

  const handleLogout = () => {
    dispatch(signOut());
  };

  const button = (
    <EuiHeaderSectionItemButton
      aria-controls={id}
      aria-expanded={isOpen}
      aria-haspopup="true"
      aria-label="Menu usuario autenticado"
      onClick={onMenuButtonClick}
    >
      <EuiAvatar size="l" name={displayName} imageUrl={photoURL} />
    </EuiHeaderSectionItemButton>
  );

  return (
    <EuiPopover
      id={id}
      ownFocus
      button={button}
      isOpen={isOpen}
      anchorPosition="downRight"
      closePopover={handleClosePopover}
      panelPaddingSize="none"
    >
      <div style={{ width: 320 }}>
        <EuiFlexGroup
          gutterSize="m"
          className="euiHeaderProfile"
          responsive={false}
        >
          <EuiFlexItem grow={false}>
            <EuiAvatar size="xl" name={displayName ?? ""} imageUrl={photoURL} />
          </EuiFlexItem>

          <EuiFlexItem>
            <EuiText>
              <p>{displayName}</p>
            </EuiText>

            <EuiSpacer size="m" />

            <EuiFlexItem>
              <EuiFlexGroup justifyContent="flexEnd">
                <EuiFlexItem grow={false}>
                  <EuiButtonEmpty
                    iconType="exit"
                    size="s"
                    onClick={handleLogout}
                    data-test-subj="logout-button"
                  >
                    Salir
                  </EuiButtonEmpty>
                </EuiFlexItem>
              </EuiFlexGroup>
            </EuiFlexItem>
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    </EuiPopover>
  );
}
