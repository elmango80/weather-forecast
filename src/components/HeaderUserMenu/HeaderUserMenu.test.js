import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

import HeaderUserMenu from "./index";

const middleware = [thunk];
const mockStore = configureStore(middleware);

const initState = {
  auth: {
    uid: "IejeJtqqpMbZQMtFVGhwBibaCDM2",
    displayName: "John Doe",
    photoURL: "http://dummyimage.com/133x128.png",
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <HeaderUserMenu />
  </Provider>
);

jest.mock("redux/actions/auth", () => ({
  signOut: jest.fn(),
}));

describe("Test <HeaderUserMenu />", () => {
  test("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch signOut", () => {
    // wrapper.find("button[data-test-subj='logout-button']").prop("onClick")();

    // expect(signOut).toHaveBeenCalled();

    console.log(wrapper.find("button[data-test-subj='logout-button']"));
  });
});
