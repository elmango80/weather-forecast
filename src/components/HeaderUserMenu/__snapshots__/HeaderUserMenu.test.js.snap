// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <HeaderUserMenu /> renders correctly 1`] = `
<div
  className="euiPopover euiPopover--anchorDownRight"
  id="i5e2d9660-5ef0-11eb-b5f6-75b89bdf7cfa"
>
  <div
    className="euiPopover__anchor"
  >
    <button
      aria-controls="i5e2d9660-5ef0-11eb-b5f6-75b89bdf7cfa"
      aria-expanded={false}
      aria-haspopup="true"
      aria-label="Menu usuario autenticado"
      className="euiHeaderSectionItem__button"
      onClick={[Function]}
      type="button"
    >
      <div
        aria-label="John Doe"
        className="euiAvatar euiAvatar--l euiAvatar--user"
        style={
          Object {
            "backgroundColor": "#c47c6c",
            "backgroundImage": "url(http://dummyimage.com/133x128.png)",
            "color": "#000000",
          }
        }
        title="John Doe"
      />
    </button>
  </div>
</div>
`;
